lxd/instance/drivers/driver_qemu.go

```diff
@@ -3277,13 +3277,33 @@ func (d *qemu) pid() (int, error) {
 	return pid, nil
 }
 
+// forceStop kills the QEMU prorcess if running, performs normal device & operation cleanup and sends stop
+// lifecycle event.
+func (d *qemu) forceStop() error {
+	pid, _ := d.pid()
+	if pid > 0 {
+		d.killQemuProcess(pid)
```

Comment:
Looking at `killQemuProcess` I'm not convinced that it's waiting logic is good enough.

You can only really `Wait` on a process which is one of your children. If LXD was restarted between the VM start time and it being killed, I don't think we'll suitably wait, resulting in us calling onStop, trying to undo the storage while qemu is still in D state trying to flush its writes.

Reply:
Yep good point I will relook at this. Thanks

Reply:
@stgraber it was actually fine because the `onStop()` function performs the correct wait procedure before starting device cleanup. However to improve readability I've put up this PR: https://github.com/lxc/lxd/pull/8978